traitcuration/traits/datasources/ols.py

```diff
@@ -45,3 +48,51 @@ def get_ontology_id(term_iri):
     if ontology_id == 'orphanet':
         ontology_id = 'ordo'
     return ontology_id
+
+
+def get_term_status(is_obsolete, ontology_id=None):
+    """
+    Takes the ontology_id of a term and a whether it is obsolete or not, and returns its calculated status.
+    'Obsolete' if the is_obsolete flag is true, 'Deleted' if no info was found in any ontology, 'Current' if its
+    ontology is EFO, and 'Needs Import' if info about the term was found in another ontology.
+    """
+    if ontology_id is None:
+        return Status.DELETED
+    if is_obsolete:
+        return Status.OBSOLETE
+    if ontology_id == 'efo':
+        return Status.CURRENT
+    return Status.NEEDS_IMPORT
+
+
+def check_awaiting_import_terms():
+    """
+    Checks if terms with the 'awaiting_import' status are now found in EFO, and if they are, assigns them as 'current'.
+    """
+    logger.info('CHECKING AWAITING IMPORT TERMS')
+    awaiting_import_terms = OntologyTerm.objects.filter(status=Status.AWAITING_IMPORT)
+    for term in awaiting_import_terms:
+        logger.info(f"Checking term {term.iri}")
+        term_info = make_ols_query(term.iri, 'efo')
+        if term_info is not None:
+            term.status = get_term_status(term_info['is_obsolete'], 'efo')
+            term.save()
+
+
+def check_term_status():
+    """
+    Queries OLS for terms with the 'awaiting_import', 'needs_import' and 'current' status, in their parent ontology,
+    to determine whether they have become obsolete or deleted.
+    """
+    logger.info('CHECKING ALL TERM STATUS')
+    terms = OntologyTerm.objects.filter(
+        status__in=[Status.AWAITING_IMPORT, Status.NEEDS_IMPORT, Status.CURRENT])
+    for term in terms:
+        logger.info(f"Checking term {term.iri}")
+        term_ontology_id = get_ontology_id(term.iri)
+        term_info = make_ols_query(term.iri, term_ontology_id)
+        if term_info is not None:
+            term.status = get_term_status(term_info['is_obsolete'], term_ontology_id)
+        else:
+            term.status = Status.DELETED
+        term.save()
```

Comment:
It's a question of personal preference, but I feel it would be better to have just one function here which would, in one cycle, go through all traits and make the necessary queries/status updates. That would make the logic more readable.

Also, I see that this currently the “current” terms are only queried in their native ontology, and only “awaiting import” are queried in EFO. But actually, we want to query _all_ terms both in their native ontology and in EFO (of course, if EFO is the native ontology, then we only want to make one query), and then make the appropriate decisions.

Reply:
> It's a question of personal preference, but I feel it would be better to have just one function here which would, in one cycle, go through all traits and make the necessary queries/status updates. That would make the logic more readable.

I am not sure I understood that, you mean we should merge the `check_awaiting_import_terms` and the `check_term_status` functions?

Reply:
Yes, exactly. Into one function which would go through all traits (regardless of their _current_ status), and then for _each_ trait do the necessary logic for queries and status updates. Something like (pseudocode):

```
def ols_update():
    for trait in all_traits:
        query efo
        query native ontology
        new_status = decide_status(current_status, efo_status, native_ontology_status)
```

Reply:
In that example, the `decide_status` doesn't have to be a separate function (and probably even shouldn't be). This could all be a monolithic function which both does the queries and then immediately decides the new status of the traits based on a bunch of branching **if** statements. The purpose of this is to translate the logic of choosing the new status as directly into code as possible, so that it's easy to verify its correctness

Reply:
Let me know if this is still not clear, I'll be happy to elaborate more

Reply:
Yes, I think this approach makes much more sense. Although I think it would be best then to separate the status calculation then, since there are going to be a bunch of if statements, and it is going to make the code much more readable.

Reply:
> Although I think it would be best then to separate the status calculation, since there are going to be a bunch of if statements, and it is going to make the code much more readable.

Well yes, actually a good point. I agree.